modules/data.atmosphere/R/download.GLDAS.R

```diff
@@ -0,0 +1,156 @@
+                   units = c('Kelvin',"W/m2","Pascal","W/m2","m/s","g/g","kg/m2/s") 
+                   )
+  
+  for (i in 1:rows){
+    year = ylist[i]    
+    
+    # figure out how many days we're working with
+    if(rows>1 & !i==1 & !i==rows){ # If we have multiple years and we're not in the first or last year, we're taking a whole year
+      nday  = ifelse(year%%4 == 0,366,365) # leap year or not; days per year
+      days.use = 1:nday
+    } else if(rows==1){
+      # if we're working with only 1 year, lets only pull what we need to
+      nday  = ifelse(year%%4 == 0,366,365) # leap year or not; days per year
+      day1 <- yday(start_date)
+      # Now we need to check whether we're ending on the right day
+      day2 <- yday(end_date)
+      days.use = day1:day2
+      nday=length(days.use) # Update nday
+    } else if(i==1) {
+      # If this is the first of many years, we only need to worry about the start date
+      nday  = ifelse(year%%4 == 0,366,365) # leap year or not; days per year
+      day1 <- yday(start_date)
+      days.use = day1:nday
+      nday=length(days.use) # Update nday
+    } else if(i==1) {
+      # If this is the last of many years, we only need to worry about the start date
+      nday  = ifelse(year%%4 == 0,366,365) # leap year or not; days per year
+      day2 <- yday(end_date)
+      days.use = 1:day2
+      nday=length(days.use) # Update nday
+    }
+    ntime = nday*8 # leap year or not*time slice (3-hourly)
+
+    loc.file = file.path(outfolder,paste("GLDAS",year,"nc",sep="."))
+    
+    ## Create dimensions
+    lat <- ncdim_def(name='latitude', units='degree_north', vals=lat.in, create_dimvar=TRUE)
+    lon <- ncdim_def(name='longitude', units='degree_east', vals=lon.in, create_dimvar=TRUE)
+    time <- ncdim_def(name='time', units="sec", vals=seq((min(days.use)*24*360), (max(days.use)+1-1/8)*24*360, length.out=ntime), create_dimvar=TRUE, unlim=TRUE)
+    dim=list(lat,lon,time)
+    
+    var.list = list()
+    dat.list = list()
+
+    # Defining our dimensions up front
+    for(j in 1:nrow(var)){
+      var.list[[j]] = ncvar_def(name=as.character(var$CF.name[j]), units=as.character(var$units[j]), dim=dim, missval=-999, verbose=verbose)
+      dat.list[[j]] <- array(NA, dim=c(length(lat.in), length(lon.in), ntime)) # Go ahead and make the arrays
+    }
+    names(var.list) <- names(dat.list) <- var$CF.name
+
+    ## get data off OpenDAP
+    for(j in 1:length(days.use)){
+      date.now <- as.Date(days.use[j], origin=as.Date(paste0(year-1,"-12-31")))
+      mo.now <- str_pad(month(date.now), 2, pad="0")
+      day.mo <- str_pad(day(date.now), 2, pad="0")
+      doy <- str_pad(days.use[j], 3, pad="0")
+      
+      # Because the suffixes are really different for these files, lets get a list and go through each day
+      dap.log <- data.frame(readHTMLTable(paste0(dap_base, "/",year, "/", doy, "/catalog.html")))
```

Comment:
It might be worth implementing at each level as we go so that we never try to query years/days that aren't there.  Might be worth also putting into NLDAS, too.  I left it alone for now because I didn't want to risk slowing things down even more.